### PR TITLE
Add script to collate and merge gtsfm reports

### PR DESCRIPTION
**Pull request by johnwlambert**
Allows side by side comparisons of tables only (not plots).

### CODE REVIEW

File: gtsfm/evaluation/merge_reports.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    return new_table_dict
+
+
+def main() -> None:
+    """ """
+
+    report_fpath1 = "/Users/jlambert/Downloads/prev_master/metrics-deep_front_end____skydio-32_______________32____jpg____gdrive_______colmap-loader____760___true.zip/gtsfm_metrics_report.html"
+
+    report_fpath2 = "/Users/jlambert/Downloads/rcc_median/metrics-deep_front_end____skydio-32_______________32____jpg____gdrive_______colmap-loader____760___true.zip/gtsfm_metrics_report.html"
```
Comment (akshay-krishnan) [inline gtsfm/evaluation/merge_reports.py:83]: Remove these, make them command line args?

File: gtsfm/evaluation/merge_reports.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    """ """  # TODO(johnwlambert): remove spaces from artifact file names
```
Comment (akshay-krishnan) [inline gtsfm/evaluation/merge_reports.py:85]: Remove comment if this is done

File: gtsfm/evaluation/merge_reports.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    return table_dict
+
+
+def print_tables_tabulate(merged_table_dict: GTSFM_MERGED_REPORT_TABLES) -> None:
+    """Dump to stdout Markdown-formatted tables."""
+    for tab_name, tab_metrics in merged_table_dict.items():
+
+        headers = [f"{tab_name}", "Report 1", "Report 2"]
+        table = tab_metrics
+        print(tabulate(table, headers, tablefmt="github"))
```
Comment (akshay-krishnan) [inline gtsfm/evaluation/merge_reports.py:60]: Might be nice to have argument to control whether we need a html or github.

**Review by akshay-krishnan — CHANGES_REQUESTED** [review]

File: gtsfm/evaluation/merge_reports.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        # add newline between tables
+        print()
+
+
+def merge_tables(table_dict1: GTSFM_REPORT_TABLES, table_dict2: GTSFM_REPORT_TABLES) -> GTSFM_MERGED_REPORT_TABLES:
+    """Combine 2-column tables from single reports, to 3-column tables, with columns corresponding to each report."""
```
Comment (womackj1) [inline gtsfm/evaluation/merge_reports.py:66]: Note: I may need to generalize this code to also combine 3-column tables to 5-column tables for elements with COLMAP comparison.

File: gtsfm/evaluation/merge_reports.py (hunk 1 above)
Comment (johnwlambert) [reply gtsfm/evaluation/merge_reports.py]: thanks, fixed now

File: gtsfm/evaluation/merge_reports.py (hunk 2 above)
Comment (johnwlambert) [reply gtsfm/evaluation/merge_reports.py]: thanks, fixed now, works without having to modify the file paths in any way

File: gtsfm/evaluation/merge_reports.py (hunk 3 above)
Comment (johnwlambert) [reply gtsfm/evaluation/merge_reports.py]: good idea, i added a command line option for that

File: gtsfm/evaluation/merge_reports.py (hunk 4 above)
Comment (johnwlambert) [reply gtsfm/evaluation/merge_reports.py:56]: Sure, feel free to add a separate PR into this branch with the changes you propose. Although for Colmap, we don't have an HTML report, so maybe a separate script is needed altogether to get the info

File: gtsfm/evaluation/merge_reports.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+Authors: John Lambert
+"""
+
+import argparse
+from collections import defaultdict
+from typing import Any, Dict, List, Tuple
+
+from tabulate import tabulate
+
+GTSFM_REPORT_TABLES = Dict[str, Dict[str, Any]]
```
Comment (ayushbaid) [inline gtsfm/evaluation/merge_reports.py:12]: I'd prefer the name explicitly convey that is is a table coming from a single report.

Also, think we can drop the GTSFM prefix

File: gtsfm/evaluation/merge_reports.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+# contains info about each metric, from two separate reports
+GTSFM_MERGED_REPORT_TABLES = Dict[str, List[Tuple[str, Any, Any]]]
```
Comment (ayushbaid) [inline gtsfm/evaluation/merge_reports.py:14]: Why does the type of the value differ? I think the only difference should be going from 2-tuple to 3-tuple

File: gtsfm/evaluation/merge_reports.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+
+def extract_tables_from_report(report_fpath: str) -> GTSFM_REPORT_TABLES:
```
Comment (ayushbaid) [inline gtsfm/evaluation/merge_reports.py:17]: Have you looked into third-party libraries to do this?

Example: pandas does this out-of-the-box. https://pbpython.com/pandas-html-table.html. We can also merge dataframes pretty easily.

File: gtsfm/evaluation/merge_reports.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    for tab_name, tab_metrics1 in table_dict1.items():
+        for tab_name1, tab_val1 in tab_metrics1.items():
+            tab_val2 = table_dict2[tab_name].get(tab_name1, None)
+            new_table_dict[tab_name] += [(tab_name1, tab_val1, tab_val2)]
+
+    return new_table_dict
+
+
+def print_tables_tabulate(merged_table_dict: GTSFM_MERGED_REPORT_TABLES, table_format: str) -> None:
```
Comment (ayushbaid) [inline gtsfm/evaluation/merge_reports.py:77]: Why not output a html table itself? I would prefer exactly the same structure of the diff-report as the original report, but with more columns and added information in the plots. 

Printing as markdown tables is not a scalable solution as we cannot do the plots in the same way.

File: gtsfm/evaluation/merge_reports.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    table_dict1 = extract_tables_from_report(report_fpath1)
+    table_dict2 = extract_tables_from_report(report_fpath2)
+    merged_table_dict = merge_tables(table_dict1, table_dict2)
+    print_tables_tabulate(merged_table_dict, output_format)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--report_fpath1", required=True, help="Path to previous report.")
+    parser.add_argument("--report_fpath2", required=True, help="Path to new report.")
```
Comment (ayushbaid) [inline gtsfm/evaluation/merge_reports.py:99]: would prefer `report1_fpath` and `report2_fpath`

**Review by ayushbaid — CHANGES_REQUESTED** [review]

File: gtsfm/evaluation/merge_reports.py (hunk 8 above)
Comment (johnwlambert) [reply gtsfm/evaluation/merge_reports.py]: @ayushbaid sure, we can do that -- dump an HTML file you mean to disk, instead of just printing to stdout?

**Review by ayushbaid — APPROVED** [review]

**Pull request merged**